subprojects/launcher/src/main/java/org/gradle/launcher/daemon/server/DaemonServices.java

```diff
@@ -80,7 +82,12 @@ public DaemonServices(DaemonServerConfiguration configuration, ServiceRegistry l
         this.loggingManager = loggingManager;
 
         addProvider(new DaemonRegistryServices(configuration.getBaseDir()));
-        addProvider(new GlobalScopeServices(true, additionalModuleClassPath));
+        addProvider(new GlobalScopeServices(true, additionalModuleClassPath, new BuildCancellationTokenFactory() {
```

Comment:
❌ I'm afraid this won't work. Please try running a build after the first one was cancelled. I think the second one will be cancelled as well even if the user didn't want to, because this token instance will be cached.

Reply:
Thanks for the catch, but it works, and current test cases already covered it:

```
    private void assertTaskIsCancellable(String task, boolean buildCacheEnabled = false) {
        startBuild(task, buildCacheEnabled)
        cancelBuild(task)

        def build = executer.withTasks("tasks").withArguments("--debug").start()
        build.waitForFinish()
        assert daemons.daemons.size() == 1
    }
```

This does a cancellation then runs a `tasks` task.

The token will be updated  every time when a new build start:

`DaemonStateCoordinator.runCommand()`
↓
`DaemonStateCoordinator.onStartCommand()`
↓
`DaemonStateCoordinator.updateCancellationToken()`

I think this is the correct way to do it, but I'm bitten by another issue:

[this method](https://github.com/gradle/gradle/blob/efb3eafbbeb610ce246425b91a168c8290910038/subprojects/core/src/main/java/org/gradle/internal/filewatch/jdk7/WatchServiceFileWatcherBacking.java#L174) is a workaround of [JDK-8011537](https://bugs.java.com/view_bug.do?bug_id=8011537) and it interrupts the thread (`DefaultExecHandle.waitForFinish`) which is waiting for forked process to exit. So I decided to revert that change.

Reply:
Ah right sorry I had missed that the daemon lookup was inside the token, not inside the provider method. All good.

Reply:
Let's try the refactoring again on `master` and merge this into `release` for now.

Reply:
Understand. Thanks!